File: Php/Infrastructure/Data/MySql/MySqlUserStorage.php
namespace Infrastructure\Data\MySql {

	use Infrastructure\Data\Interfaces\UserStorage;
	use Infrastructure\Data\MySql\StorageProvider;
	use Domain\Classes\User;
	use app\User\UserFactory;

	class MySqlUserStorage implements UserStorage {

		private $db;
		private $userFactory;

		public function __construct(StorageProvider $db,
									UserFactory $userFactory){
			db = $db;
			userFactory = $userFactory;
		}

		public static function className(){
			return __CLASS__;
		}

		public function getUserByEmail($email){
			$sql =	getUserBaseSql();
			$sql .=	"WHERE u.Email = '{$email}'";
			return getUser($sql);
		}

		public function getUserByName($userName){
			$sql =	getUserBaseSql();
			$sql .=	"WHERE u.UserName = '{$userName}'";
			return getUser($sql);
		}

		public function getUserByToken($token){
			$sql =	getUserBaseSql();
			$sql .=	"WHERE u.Token = '{$token}'";
			return getUser($sql);
		}

		public function getUserByCredentials($userNameOrEmail, $password){
			$sql = getUserBaseSql();
			$sql .= "WHERE (u.UserName = '{$userNameOrEmail}' OR u.Email = '{$userNameOrEmail}') " .
					"AND u.Password = '{$password}'";
			return getUser($sql);
		}

		private function getUserBaseSql(){
			$sql =	"SELECT u.UserID, u.UserName, u.DisplayName, u.RealName, u.Email, u.Token, u.Password, u.Salt, u.RoleID " .
					"FROM user u ";
			return $sql;
		}

		private function getUser($sql){
			$res = db.query($sql);
			foreach($res.fetchAll() as $row){
				return userFromDbRow($row);
			}
			return null;
		}

		public function getUsers(){
			$sql =	getUserBaseSql();
			$sql .=	"ORDER BY u.DisplayName";
			$res = db.query($sql);
			$users = array();
			foreach($res.fetchAll() as $row){
				$users[] = userFromDbRow($row);
			}
			return $users;
		}

		/**
		 * @param User $user
		 * @return bool
		 */
		public function updateUser($user){
			$sql =	"UPDATE user u " .
					"SET " .
						"DisplayName = '{$user.getDisplayName()}', " .
						"RealName = '{$user.getRealName()}', " .
						"Email = '{$user.getEmail()}' " .
					"WHERE UserID = {$user.getId()}";
			$rowCount = db.execute($sql);
			return $rowCount > 0;
		}

		/**
		 * @param User $user
		 * @return int
		 */
		public function addUser(User $user){
			$sql =	"INSERT INTO user " .
					"(UserName, DisplayName, Email) " .
					"VALUES " .
					"('{$user.getUserName()}', '{$user.getDisplayName()}', '{$user.getEmail()}')";
			$rowCount = db.execute($sql);
			return db.getLatestInsertId($rowCount > 0);
		}

		public function deleteUser(User $user){
			$sql =	"DELETE FROM user u " .
					"WHERE UserID = {$user.getId()}";
			$rowCount = db.execute($sql);
			return $rowCount > 0;
		}

		public function getSalt($userNameOrEmail){
			$sql =	"SELECT u.Salt " .
					"FROM user u " .
					"WHERE (u.UserName = '{$userNameOrEmail}' OR u.Email = '{$userNameOrEmail}')";
			$res = db.query($sql);
			$salt = "";
			foreach($res.fetchAll() as $row){
				$salt = $row["Salt"];
				break;
			}
			return $salt;
		}

		public function setSalt(User $user, $salt){
			$sql =	"UPDATE user u " .
				"SET " .
					"Salt = '{$salt}' " .
				"WHERE UserName = '{$user.getUserName()}'";
			$rowCount = db.execute($sql);
			return $rowCount > 0;
		}

		public function setEncryptedPassword(User $user, $encryptedPassword){
			$sql =	"UPDATE user u " .
				"SET " .
					"Password = '{$encryptedPassword}' " .
				"WHERE UserName = '{$user.getUserName()}'";
			$rowCount = db.execute($sql);
			return $rowCount > 0;
		}

		public function setToken(User $user, $token){
			$sql =	"UPDATE user u " .
				"SET " .
					"Token = '{$token}' " .
				"WHERE UserName = '{$user.getUserName()}'";
			$rowCount = db.execute($sql);
			return $rowCount > 0;
		}

		public function getToken(User $user){
			$sql =	"SELECT u.Token " .
				"FROM user u " .
				"WHERE u.UserName = '{$user.getUserName()}'";
			$res = db.query($sql);
			$token = null;
			foreach($res.fetchAll() as $row){
				$token = $row["Token"];
				break;
			}
			return $token;
		}

		public function userFromDbRow($row){
			$id = $row["UserID"];
			$userName = $row["UserName"];
			$displayName = $row["DisplayName"];
			$realName = $row["RealName"];
			$email = $row["Email"];
			$globalRole = $row["RoleID"];
			$user = userFactory.createUser($id, $userName, $displayName, $realName, $email, $globalRole);
			return $user;
		}

	}

}

File: Infrastructure/Data/Storage/MySqlUserStorage.cs
using System.Collections.Generic;
using System.Data;
using Core.Classes;
using Infrastructure.Data.Storage.Interfaces;
using Infrastructure.Factories;
using MySql.Data.MySqlClient;

namespace Infrastructure.Data.Storage {
    public class MySqlUserStorage : IUserStorage 
    {
	    private readonly IStorageProvider _storageProvider;
	    private readonly IUserFactory _userFactory;

	    public MySqlUserStorage(IStorageProvider storageProvider, IUserFactory userFactory)
	    {
	        _storageProvider = storageProvider;
	        _userFactory = userFactory;
	    }

		public User GetUserByEmail(string email){
			var sql = GetUserBaseSql();
			sql += "WHERE u.Email = '{0}'";
		    sql = string.Format(sql, email);
			return getUser(sql);
		}

		public User GetUserByName(string userName){
			var sql = GetUserBaseSql();
			sql += "WHERE u.UserName = '{0}'";
            sql = string.Format(sql, userName);
			return getUser(sql);
		}

		public User GetUserByToken(string token){
			var sql =	GetUserBaseSql();
			sql +=	"WHERE u.Token = '{0}'";
			sql = string.Format(sql, token);
            return getUser(sql);
		}

		public User GetUserByCredentials(string userNameOrEmail, string password){
			var sql = GetUserBaseSql();
			sql += "WHERE (u.UserName = '{0}' OR u.Email = '{0}') AND u.Password = '{1}'";
			sql = string.Format(sql, userNameOrEmail, password);
            return getUser(sql);
		}

		private string GetUserBaseSql(){
			return "SELECT u.UserID, u.UserName, u.DisplayName, u.RealName, u.Email, u.Token, u.Password, u.Salt, u.RoleID FROM user u ";
		}

		private User getUser(string sql){
            var reader = _storageProvider.Query(sql);
            while (reader.Read())
            {
                return UserFromDbRow(reader);
            }
			return null;
		}

		public List<User> GetUsers(){
			var sql =	GetUserBaseSql();
			sql +=	"ORDER BY u.DisplayName";
			var reader = _storageProvider.Query(sql);
			var users = new List<User>();
            while (reader.Read())
            {
                users.Add(UserFromDbRow(reader));
            }
			return users;
		}

		public bool UpdateUser(User user){
			var sql =	"UPDATE user u " +
					"SET " +
						"DisplayName = '{0}', " +
						"RealName = '{1}', " +
						"Email = '{2}' " +
					"WHERE UserID = {3}";
		    sql = string.Format(sql, user.DisplayName, user.RealName, user.Email, user.Id);
		    var rowCount = _storageProvider.Execute(sql);
			return rowCount > 0;
		}

		public int AddUser(User user){
			var sql =	"INSERT INTO user " +
					"(UserName, DisplayName, Email) " +
					"VALUES " +
					"('{0}', '{1}', '{2}')";
            sql = string.Format(sql, user.UserName, user.DisplayName, user.Email);
            var id = _storageProvider.ExecuteInsert(sql);
			return id;
		}

		public bool DeleteUser(User user){
			var sql =	"DELETE FROM user u " +
					"WHERE UserID = {0}";
            sql = string.Format(sql, user.Id);
			var rowCount = _storageProvider.Execute(sql);
			return rowCount > 0;
		}

		public string GetSalt(string userNameOrEmail){
			var sql =	"SELECT u.Salt " +
					"FROM user u " +
					"WHERE (u.UserName = '{0}' OR u.Email = '{0}')";
            sql = string.Format(sql, userNameOrEmail);
            var reader = _storageProvider.Query(sql);
			while(reader.Read()){
				return reader.GetString("Salt");
			}
			return "";
		}

		public bool SetSalt(User user, string salt){
			var sql =	"UPDATE user u " +
				"SET " +
					"Salt = '{0}' " +
				"WHERE UserName = '{1}'";
		    sql = string.Format(sql, salt, user.UserName);
			var rowCount = _storageProvider.Execute(sql);
			return rowCount > 0;
		}

		public bool SetEncryptedPassword(User user, string encryptedPassword){
			var sql =	"UPDATE user u " +
				"SET " +
					"Password = '{0}' " +
				"WHERE UserName = '{1}'";
			sql = string.Format(sql, encryptedPassword, user.UserName);
            var rowCount = _storageProvider.Execute(sql);
			return rowCount > 0;
		}

		public bool SetToken(User user, string token){
			var sql =	"UPDATE user u " +
				"SET " +
					"Token = '{0}' " +
				"WHERE UserName = '{1}'";
            sql = string.Format(sql, token, user.UserName);
			var rowCount = _storageProvider.Execute(sql);
			return rowCount > 0;
		}

		public string GetToken(User user){
			var sql =	"SELECT u.Token " +
				"FROM user u " +
				"WHERE u.UserName = '{0}'";
            sql = string.Format(sql, user.UserName);
			var reader = _storageProvider.Query(sql);
			while(reader.Read())
			{
			    return reader.GetString("Token");
			}
			return null;
		}

        private User UserFromDbRow(StorageDataReader reader){
			var id = reader.GetInt("UserID");
            var userName = reader.GetString("UserName");
			var displayName = reader.GetString("DisplayName");
			var realName = reader.GetString("RealName");
			var email = reader.GetString("Email");
			var globalRole = (Role)reader.GetInt("RoleID");
			return _userFactory.Create(id, userName, displayName, realName, email, globalRole);
		}

	}

}